src/main/php/unittest/TestRun.class.php

```diff
@@ -0,0 +1,323 @@
+   * Notify listeners
+   *
+   * @param  string $method
+   * @param  var[] $args
+   * @return void
+   */
+  private function notify($method, $args) {
+    foreach ($this->listeners as $l) {
+      $l->{$method}(...$args);
+    }
+  }
+
+  /**
+   * Returns values
+   *
+   * @param  unittest.TestCase test
+   * @param  var annotation
+   * @return var values a traversable structure
+   */
+  private function valuesFor($test, $annotation) {
+    if (!is_array($annotation)) {               // values("source")
+      $source= $annotation;
+      $args= [];
+    } else if (isset($annotation['source'])) {  // values(source= "src" [, args= ...])
+      $source= $annotation['source'];
+      $args= isset($annotation['args']) ? $annotation['args'] : [];
+    } else {                                    // values([1, 2, 3])
+      return $annotation;
+    }
+
+    // Route "ClassName::methodName" -> static method of the given class,
+    // "self::method" -> static method of the test class, and "method" 
+    // -> the run test's instance method
+    if (false === ($p= strpos($source, '::'))) {
+      return typeof($test)->getMethod($source)->setAccessible(true)->invoke($test, $args);
+    }
+
+    $ref= substr($source, 0, $p);
+    if ('self' === $ref) {
+      $class= typeof($test);
+    } else if (strstr($ref, '.')) {
+      $class= XPClass::forName($ref);
+    } else {
+      $class= new XPClass($ref);
+    }
+    return $class->getMethod(substr($source, $p+ 2))->invoke(null, $args);
+  }
+
+  /**
+   * Returns values
+   *
+   * @param  var $annotatable
+   * @return var[]
+   */
+  private function actionsFor($annotatable) {
+    $r= [];
+    if ($annotatable->hasAnnotation('action')) {
+      $action= $annotatable->getAnnotation('action');
+      if (is_array($action)) {
+        foreach ($action as $a) {
+          if ($a instanceof TestAction) $r[]= $a;
+        }
+      } else {
+        if ($action instanceof TestAction) $r[]= $action;
+      }
+    }
+    return $r;
+  }
+
+  /**
+   * Invoke a block, wrap PHP5 and PHP7 native base exceptions in lang.Error
+   *
+   * @param  function(?): void $block
+   * @param  var $arg
+   * @return void
+   */
+  private function invoke($block, $arg) {
+    try {
+      $block($arg);
+    } catch (Throwable $e) {
+      throw $e;
+    } catch (\Exception $e) {
+      throw Throwable::wrap($e);
+    } catch (\Throwable $e) {
+      throw Throwable::wrap($e);
+    }
+  }
+
+  /**
+   * Record outcome, notifying listeners
+   *
+   * @param  string $type
+   * @param  unittest.TestOutcome $result
+   * @return void
+   */
+  private function record($type, $outcome) {
+    $this->notify($type, [$this->result->record($outcome)]);
+    Errors::clear();
+  }
+
+  /**
+   * Run a test case.
+   *
+   * @param  unittest.TestCase $test
+   * @param  unittest.TestResult $result
+   * @return void
+   */
+  private function run($test) {
+    $class= typeof($test);
+    $method= $class->getMethod($test->name);
```

Comment:
Idea for future refactoring: Abstract these into something similar to `TestGroup`.